app/src/main/java/com/github/libretube/services/DownloadService.kt

```diff
@@ -1,182 +1,524 @@
+        val fileName = intent.getStringExtra(IntentData.fileName) ?: videoId
+        val videoFormat = intent.getStringExtra(IntentData.videoFormat)
+        val videoQuality = intent.getStringExtra(IntentData.videoQuality)
+        val audioFormat = intent.getStringExtra(IntentData.audioFormat)
+        val audioQuality = intent.getStringExtra(IntentData.audioQuality)
+        val subtitleCode = intent.getStringExtra(IntentData.subtitleCode)
+
+        scope.launch {
+            try {
+                val streams = RetrofitInstance.api.getStreams(videoId)
+
+                awaitQuery {
+                    Database.downloadDao().insertDownload(
+                        Download(
+                            videoId = videoId,
+                            title = streams.title ?: "",
+                            thumbnailPath = File(
+                                DownloadHelper.getDownloadDir(
+                                    this@DownloadService,
+                                    DownloadHelper.THUMBNAIL_DIR
+                                ),
+                                fileName
+                            ).absolutePath,
+                            description = streams.description ?: "",
+                            uploadDate = streams.uploadDate,
+                            uploader = streams.uploader ?: ""
+                        )
+                    )
+                }
+                streams.thumbnailUrl?.let { url ->
+                    ImageHelper.downloadImage(
+                        this@DownloadService,
+                        url,
+                        File(
+                            DownloadHelper.getDownloadDir(
+                                this@DownloadService,
+                                DownloadHelper.THUMBNAIL_DIR
+                            ),
+                            fileName
+                        ).absolutePath
+                    )
+                }
+
+                val downloadItems = streams.toDownloadItems(
+                    videoId,
+                    fileName,
+                    videoFormat,
+                    videoQuality,
+                    audioFormat,
+                    audioQuality,
+                    subtitleCode
+                )
+                downloadItems.forEach { start(it) }
+            } catch (e: Exception) {
+                return@launch
+            }
         }
 
-        return super.onStartCommand(intent, flags, startId)
+        return START_NOT_STICKY
     }
 
-    override fun onBind(intent: Intent?): IBinder? {
-        TODO("Not yet implemented")
+    /**
+     * Initiate download [Job] using [DownloadItem] by creating file according to [FileType]
+     * for the requested file.
+     */
+    private fun start(item: DownloadItem) {
+        val file: File = when (item.type) {
+            FileType.AUDIO -> {
+                val audioDownloadDir = DownloadHelper.getDownloadDir(
```

Comment:
That (and the two following statements for videos and subtitles) should be using the above mentioned function as well.